pkg/thethingsgateway/cups/handlers.go

```diff
@@ -0,0 +1,161 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cups
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/gogo/protobuf/types"
+	echo "github.com/labstack/echo/v4"
+	"go.thethings.network/lorawan-stack/pkg/pfconfig/shared"
+	"go.thethings.network/lorawan-stack/pkg/rpcmetadata"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+// gatewayInfoResponse represents a minimal gateway info response for The Things Gateway.
+type gatewayInfoResponse struct {
+	FrequencyPlanID  string `json:"frequency_plan"`
+	FrequencyPlanURL string `json:"frequency_plan_url"`
+	FirmwareURL      string `json:"firmware_url"`
+	Router           struct {
+		MQTTAddress string `json:"mqtt_address"`
+	} `json:"router"`
+	AutoUpdate bool `json:"auto_update"`
+}
+
+func (s *Server) handleGatewayInfo(c echo.Context) error {
+	ctx := s.getContext(c)
+	gatewayIDs := c.Get(gatewayIDKey).(ttnpb.GatewayIdentifiers)
+	if gateway, err := s.getRegistry(ctx, &gatewayIDs).Get(ctx, &ttnpb.GetGatewayRequest{
+		GatewayIdentifiers: gatewayIDs,
+		FieldMask: types.FieldMask{Paths: []string{
+			"frequency_plan_id",
+			"gateway_server_address",
+			"update_channel",
+			"auto_update",
+		}},
+	}, s.getAuth(c)); err != nil {
+		return err
+	} else {
+		freqPlanURL := &url.URL{
+			Scheme: c.Scheme(),
+			Host:   c.Request().Host,
```

Comment:
What we can do to work around https://github.com/TheThingsProducts/gateway/issues/60 is fill the stack's default ports for Scheme into Host in case the Host of the request doesn't contain a port and there is no indication of being behind a reverse proxy (perhaps by looking at the `X-Forwarded-For` header).

That way, if the stack is configured for 1885/8885, then the ports are added. If the stack is configured for 80/443 we don't break anything by unnecessarily adding the ports. I think we can just assume that the default http(s) ports are used if the stack is behind a reverse proxy.